edge-modules/azure-monitor/src/CertificateGenerator/CertGenerator.cs

```diff
@@ -0,0 +1,287 @@
+                try
+                {
+                    RegisterWithOms(
+                       cert, AgentGuid, logAnalyticsWorkspaceId, logAnalyticsWorkspaceKey, logAnalyticsWorkspaceDomain);
+
+                    // Return or break.
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    currentRetry++;
+
+                    // Check if the exception thrown was a transient exception
+                    // based on the logic in the error detection strategy.
+                    // Determine whether to retry the operation, as well as how
+                    // long to wait, based on the retry strategy.
+                    if (currentRetry > 3)
+                    {
+                        // If this isn't a transient error or we shouldn't retry,
+                        // rethrow the exception.
+                        Console.WriteLine("exception occurred : {0}", ex.Message);
+                        throw;
+                    }
+                }
+
+                // Wait to retry the operation.
+                // Consider calculating an exponential delay here and
+                // using a strategy best suited for the operation and fault.
+                Task.Delay(1000);
+            }
+        }
+
+        public static (X509Certificate2 tempCert, (string, byte[]), string) RegisterAgentWithOMS(string logAnalyticsWorkspaceId, string logAnalyticsWorkspaceKey, string logAnalyticsWorkspaceDomain)
+        {
+            X509Certificate2 agentCert = null;
+            string certString;
+            byte[] certBuf;
+            string keyString;
+
+            var agentGuid = Guid.NewGuid().ToString("B");
+
+            try
+            {
+                Environment.SetEnvironmentVariable("CI_AGENT_GUID", agentGuid);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Failed to set env variable (CI_AGENT_GUID)" + ex.Message);
+            }
+
+            try
+            {
+                (agentCert, (certString, certBuf), keyString) = CreateSelfSignedCertificate(agentGuid, logAnalyticsWorkspaceId);
+
+                if (agentCert == null)
+                {
+                    throw new Exception($"creating self-signed certificate failed for agentGuid : {agentGuid} and workspace: {logAnalyticsWorkspaceId}");
+                }
+
+                Console.WriteLine($"Successfully created self-signed certificate  for agentGuid : {agentGuid} and workspace: {logAnalyticsWorkspaceId}");
+
+                RegisterWithOmsWithBasicRetryAsync(agentCert, agentGuid,
+                    logAnalyticsWorkspaceId,
+                    logAnalyticsWorkspaceKey,
+                    logAnalyticsWorkspaceDomain);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Registering agent with OMS failed (are the Log Analytics Workspace ID and Key correct?) : {0}", ex.Message);
```

Comment:
Please expand the `OMS` into the full name.

Reply:
@yophilav We are just checking the code in from the Monitor team and will be making these types of code changes in separate PRs.